### PR TITLE
tls: reduce the security strength of the CCM8 cipher suites

### LINKED ISSUE #16154: Reconsider strength of CCM_8 TLS ciphers
As far as I can tell, we don't give any particular consideration to the truncated authentication tag of the CCM_8 ciphers when we're considering cipher strength and security level.  E.g., at `openssl ciphers -s -v 'ALL@SECLEVEL=2' we still see ECDHE-ECDSA-AES256-CCM8, AES128-CCM8, etc., and I think that TLS_AES_128_CCM_8_SHA256 is also treated as being strong.
It's not entirely clear that we should be treating ciphers that only have a 64-bit authentication tag as providing that level of security.

(Inspired by https://github.com/openssl/openssl/issues/13689#issuecomment-752220290)

### PR DESCRIPTION
**Pull request by paulidale**
Alternative to #16160

CCM8 suites have a 64 bit authentication tag which is pretty short.

The changes reduces these cipher suites to security levels below 2.

Fixes #16154

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by kaduk** [comment]
Is there a mechanism behind picking "80" or just expedience for having it produce the right outcome?
(The geometric mean of 64 and 128 is roughly 90, FWIW...)

**Comment by slontis** [comment]
> Is there a mechanism behind picking "80" or just expedience for having it produce the right outcome?
> (The geometric mean of 64 and 128 is roughly 90, FWIW...)

ssl_security_default_callback() calls ssl_get_security_level_bits() which has a  minbits table.

**Comment by paulidale** [comment]
I picked the value we use for everything else that is security level 1.  As far as I can tell, we only use 80, 112, 128, 192 and 256 for strengths.  I didn't check too hard.

I've no objection to making it 90.  I don't know if any code depends on it being one of the standard values.  I'd be quite surprised if somebody somewhere didn't write this kind of rubbish:
```c
switch (strength) {
case 80: ... break;
case 112: ... break;
case 128: ... break;
case 192: ... break;
case 256: ... break;
}
```

**Comment by paulidale** [comment]
I prefer the solution in #16160 to be honest.  But this is more data driven.

**Comment by kaduk** [comment]
> I prefer the solution in #16160 to be honest. But this is more data driven.

On first look I do, as well.  But I cannot claim to have spent the time to properly review either approach.

**Review by t8m — APPROVED** [review]
IMO this is better than the alternative.

File: ssl/s3_lib.c [hunk 1]
```diff
@@ -108,9 +108,9 @@ static SSL_CIPHER tls13_ciphers[] = {
         SSL_AEAD,
         TLS1_3_VERSION, TLS1_3_VERSION,
         0, 0,
-        SSL_NOT_DEFAULT | SSL_HIGH,
+        SSL_NOT_DEFAULT | SSL_MEDIUM,
         SSL_HANDSHAKE_MAC_SHA256,
-        128,
+        80,
```
Comment (mattcaswell) [inline ssl/s3_lib.c:113]: Should we add a short comment explaining why it is 80 and not the 128 that you might expect, e.g.

````
/* CCM8 uses a short tag, so we have a lower security strength */
````
Or something like that. Ditto for the other instances.

File: doc/man3/SSL_CTX_set_security_level.pod [hunk 2]
```diff
@@ -83,8 +83,9 @@ using MD5 for the MAC is also prohibited.
 
 Security level set to 112 bits of security. As a result RSA, DSA and DH keys
 shorter than 2048 bits and ECC keys shorter than 224 bits are prohibited.
-In addition to the level 1 exclusions any cipher suite using RC4 is also
-prohibited. SSL version 3 is also not allowed. Compression is disabled.
+In addition to the level 1 exclusions any cipher suite using RC4 and any
+cipher suite using CCM with a 64 bit authentication tag are also prohibited.
+SSL version 3 is also not allowed. Compression is disabled.
```
Comment (mattcaswell) [inline doc/man3/SSL_CTX_set_security_level.pod:88]: This probably deserves a mention in CHANGES.md

File: ssl/s3_lib.c (hunk 1 above)
Comment (paulidale) [reply ssl/s3_lib.c]: Added.

File: doc/man3/SSL_CTX_set_security_level.pod (hunk 2 above)
Comment (paulidale) [reply doc/man3/SSL_CTX_set_security_level.pod:88]: Added

**Comment by paulidale** [comment]
I've also pushed a new commit changing the strength to the more accurate 90 bits.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
> I've also pushed a new commit changing the strength to the more accurate 90 bits.
> Happy to drop this part if it is controversial.

Why is 90 bits strength more accurate?

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
@kaduk, do you have a reference for the strength?

Using the geometric mean, AES-256-CCM8 would have a strength > 112 bits and would stay at security level 2.

**Comment by kaduk** [comment]
"geometric mean" was just an attempt to backsolve how one might justify 80 bits on its own right (as opposed to just happening to trigger the desired threshold based on the current logic in the other parts of the code).
I don't think I have a reference for CCM8 strength that would give an answer larger than 64 bits.
https://datatracker.ietf.org/doc/draft-irtf-cfrg-aead-limits/ is in some senses even less optimistic about CCM8...

**Comment by mattcaswell** [comment]
One possibility is to just put it at 64 bits which is at least justifiable. Although it would mean it goes into security level 0 instead of 1. That would have meant that it is not available in the default security level any more - except for the fact that we just voted to increase the default security level to 2 anyway - so either way it won't be available in the default security level. So the distinction between 0 and 1 might not matter that much anyway.

**Comment by paulidale** [comment]
I've switched them to 80 bits again.  Going to 64 bits breaks tests.  I'll investigate and reduce the strength if I can fix things.

**Comment by paulidale** [comment]
I've lowered the security level to 64 bits and fixed the test failures.
Re-review required.

File: CHANGES.md [hunk 3]
```diff
@@ -24,6 +24,11 @@ OpenSSL 3.1
 
 ### Changes between 3.0 and 3.1 [xx XXX xxxx]
 
+ * CCM8 cipher suites in TLS have been downgraded to security level 1 because
```
Comment (kaduk) [inline CHANGES.md:27]: 0 now?

File: CHANGES.md [hunk 4]
```diff
@@ -24,6 +24,11 @@ OpenSSL 3.1
...
+   they use a short tag which lowers their strength.
```
Comment (kaduk) [inline CHANGES.md:28]: I forget if we like "tag" vs "authentication tag"

File: doc/man3/SSL_CTX_set_security_level.pod [hunk 5]
```diff
@@ -83,8 +83,9 @@ using MD5 for the MAC is also prohibited.
 
 Security level set to 112 bits of security. As a result RSA, DSA and DH keys
 shorter than 2048 bits and ECC keys shorter than 224 bits are prohibited.
-In addition to the level 1 exclusions any cipher suite using RC4 is also
-prohibited. SSL version 3 is also not allowed. Compression is disabled.
+In addition to the level 1 exclusions any cipher suite using RC4 and any
```
Comment (kaduk) [inline doc/man3/SSL_CTX_set_security_level.pod:86]: Should be in the level 1 section now and not mention "in addition to the l1 exclusions"?

**Review by kaduk — COMMENTED** [review]
This is looking more attractive now that we are using a number we can justify based on the math

File: CHANGES.md (hunk 3 above)
Comment (paulidale) [reply CHANGES.md:27]: Oops :(

File: CHANGES.md (hunk 4 above)
Comment (paulidale) [reply CHANGES.md:28]: I doubt it matters but I'll change to use _authentication tag_.

**Comment by paulidale** [comment]
Comments addressed.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the feedback and discussion.

**Pull request closed without merging**